opencog/atoms/distvalue/DistributionalValue.cc

```diff
@@ -0,0 +1,455 @@
+	}
+	return max;
+}
+
+std::vector<double> DistributionalValue::get_mode() const
+{
+	std::vector<double> probs;
+	for (auto elem : _value)
+	{
+		probs.push_back(get_mode_for(elem.second));
+	}
+	return probs;
+}
+
+double DistributionalValue::get_mode_for(double ai) const
+{
+	return (ai - 1) / (total_count() - _value.size());
+}
+
+std::vector<double> DistributionalValue::get_mean() const
+{
+	std::vector<double> probs;
+	for (auto elem : _value)
+	{
+		probs.push_back(get_mean_for(elem.second));
+	}
+	return probs;
+}
+
+double DistributionalValue::get_mean_for(double ai) const
+{
+	double count = total_count();
+	if (count == 0)
+		return 0;
+	else
+		return ai / total_count();
+}
+
+//Condenses the DV into a single strenght value
+//This should be avoided where ever possible as it loses a lot of information
+double DistributionalValue::get_fstord_mean() const
+{
+	double res = 0;
+	for (auto elem : _value)
+	{
+		//TODO: Is mode correct or should i use mean?
+		DVec mof = middle_of_interval(elem.first);
+		double mode = std::accumulate(mof.begin(),mof.end(),0.0)/mof.size();
+		res = res + mode * get_mode_for(elem.second);
+	}
+	return res;
+}
+
+//Calculates the Middle of all intervals in a Key
+DVec DistributionalValue::middle_of_interval(const DVKey &k) const
+{
+	DVec res;
+	for (auto interval : k) {
+		if (interval.size() == 1)
+			res.push_back(interval[0]);
+		res.push_back((interval[0] + interval[1])/2);
+	}
+	return res;
+}
+
+//Get the variance for all Keys
+std::vector<double> DistributionalValue::get_var() const
+{
+	std::vector<double> probs;
+	for (auto elem : _value)
+	{
+		probs.push_back(get_var_for(elem.second));
+	}
+	return probs;
+}
+
+//Get the variance for a Count
+double DistributionalValue::get_var_for(double ai) const
+{
+		double a0 = total_count();
+		return ai*(a0-ai) / a0*a0*(a0+1);
+}
+
+double DistributionalValue::total_count() const
+{
+	return _value.total_count();
+}
+
+double DistributionalValue::get_confidence() const
+{
+	int c = total_count();
+	return to_conf(c);
+}
+
+double DistributionalValue::to_conf(int c)
+{
+	return c / (c + DEFAULT_K);
+}
+
+int DistributionalValue::to_count(double cf)
+{
+	return (cf * DEFAULT_K / (1 - cf));
+}
+
+bool DistributionalValue::has_key(const DVKey &k) const
+{
+	auto it = _value.find(k);
+	return it != _value.end();
+}
+
+DVKeySeq DistributionalValue::get_keys() const
+{
+	DVKeySeq res;
+	for (auto k : _value)
+		res.push_back(k.first);
+	return res;
+}
+
+double DistributionalValue::get_count(const DVKey &h) const
+{
+	auto pos = _value.find(h);
+	if (pos != _value.end())
+		return pos->second;
+	throw RuntimeException(TRACE_INFO, "No Key for this value.");
+}
+
+//Find out how much of Key1 is contained in Key2
+double DistributionalValue::key_contained(const DVKey &ks1,const DVKey &ks2)
+{
+	Interval k1,k2;
+	//Start with the assumption that 100% of Key1 is in Key2
+	double sum = 1;
+	for (auto zipped : boost::combine(ks1,ks2))
+	{
+		boost::tie(k1,k2) = zipped;
```

Comment:
Wow, that `tie` function is really bad ass! Actually it seems to be in the standard https://en.cppreference.com/w/cpp/utility/tuple/tie and I would be bet it is supported by gcc 5.0 so if you can replace by `std::tie` it would be better.

Reply:
What's more, it is optimized away https://stackoverflow.com/questions/30212307/does-gcc-optimize-stdtie-used-only-for-readability and works for pairs https://stackoverflow.com/questions/31272966/does-there-exist-something-like-stdtie-for-stdpair#31273053 this thing :-)

Reply:
Hm, not as straightforward as I hoped. boost::combine uses boost::tuples so std::tie doesn't work for that. So I also need to replace boost::combine and there is no combine in std yet.
So I could leave as is or maybe switch to using 2 independent iterators or write my own combine function that uses std::tuple. I am leaning towards option 2 as that is also what I have done in some other places. Option 3 seems a bit overkill for this one case but if it were used in more places it might be worth it.

Reply:
Do what you feel is right, keeping `boost::tie` isn't a big deal if necessary.